CONTRIBUTING.md

```diff
@@ -800,45 +744,280 @@ Try restricting ifdefs in the header files, or in the beginning of
+You can now make your changes. Build and test them when done:
 
 ```
-$ git branch
-  3.10.x
-  3.11.x
-  3.7.x
-* ENT-3329
-  master
+$ ./autogen.sh --enable-debug && make -j4 && (cd tests/unit/ && make check)
 ```
 
+Running all the tests is quite slow, because of many acceptance tests.
+Locally, it is usually sufficient to run unit tests, and select acceptance
+tests. Then, commit and push normally:
+
 ```
-$ git pull --rebase upstream master
-From https://github.com/cfengine/masterfiles
- * branch                master     -> FETCH_HEAD
-First, rewinding head to replay your work on top of it...
-Applying: ENT-3329: Allow hubs to collect from themselves over loopback
+$ git add -p
+$ git commit
+$ git push
 ```
 
-Once you have rebased your changes on the latest changes from the upstream branch you need to force push (because history has been re-written) your branch to your fork in order to update the pull request.
+The default behavior of `git push` may or may not work for you, you can also
+specify where to push manually:
 
 ```
-$ git push origin master --force
+$ git push --set-upstream origin mytopic
 ```
 
+On your fork at Github you should now see an additional branch `mytopic`.
+This is what you use for the pull request there.
+Pushing more commits or force pushing changed commits to the same branch will
+automatically update an open Pull Request which uses that branch.
+
+
+### Cleaning up the commit history - rebasing and squashing
+
 
-#### How to rebase and squash commits
+#### Resolving conflicts using rebase
 
-[This post has a good explanation.](https://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit)
+After you branched from `upstream/master` it is possible that someone else
+(or even you) commit conflicting changes to the `master` branch.
+To resolve these conflicts in your PR branch, use:
+
+```
+$ git fetch --all
+$ git rebase upstream/master
+```
+
+This will stop at the conflicting commit, and give you instructions to
+proceed. You will have to resolve conflicts and then run:
+
+```
+$ git rebase --continue
+```
+
+This might happen multiple times (1 for each commit).
+Once done, you will have to force push your changes to the PR branch:
+
+```
+$ git push -f
+```
+
+
+#### Interactive rebase
+
+We want the commit history to be as easy to read and understand as possible.
+Sometimes we make more commits than necessary, and we want to squash them
+(combine them). A good tool for this is _interactive rebase_.
+
+On your PR branch, run:
+
+```
+$ git rebase -i upstream/master
+```
+
+This will open an editor (usually `vi` or `emacs`), with a list of your
+commits:
+
+```
+pick ac07e6d27 Added new policy function isrealdir()
+pick 001bd420b Fixed mistake in isrealdir()
+
+# Commented out instructions ....
+```
+
+The commits / commands will be applied to `upstream/master` from top to bottom
```

Comment:
I don't think `git rebase -i upstream/master` is really a good example. It mixes too many things. `git rebase -i HEAD~N` or something similar is much easier to understand, AFAICT.

For example the above *"commits/commands will be applied to `upstream/master`"* is not quite true, right? They are applied to the local branch which is being rebased to `upstream/master`.

Reply:
> I don't think `git rebase -i upstream/master` is really a good example. It mixes too many things. `git rebase -i HEAD~N` or something similar is much easier to understand, AFAICT.

I disagree. I only mentioned HEAD in one other place, and I was considering to not do it. `HEAD~N` is even more confusing to me, then you have to explain what HEAD means, what `~` means, etc. `upstream/master` is discussed in other places, and it also has another advantage: you can use the same command in (more than) 3 different situations:

1) Your PR is out of date and you need to bring it up to date with master (upstream)
2) Bringing local master branch up to date (with upstream)
3) Adding `-i` option to interactively rebase (for squashing commits etc.)

Yes, teaching and explaining these things is difficult.

> For example the above "commits/commands will be applied to upstream/master" is not quite true, right? They are applied to the local branch which is being rebased to upstream/master.

This depends on how you interpret `apply`. I think it's right, as it says there, but I also understand that it can be confusing. Wording these things correctly and in an understandable way is hard. Something like `upstream/master is used as a base (starting point), then the commits in the list are applied (without altering upstream/master)` is how I read it.

Reply:
@vpodzime Also, we should assume people to some extent understand git. I think of this as some sort of reference / command lookup. If you don't understand how rebase works or exactly what `upstream/master` you can google that.

Reply:
I updated the wording to try to be a little more clear.